actions/v3/upload_sessions.go

```diff
@@ -219,3 +264,51 @@ func sendBetaWithUploadRequest(req uploadSessionCreateReqV3) (uploadSessionCreat
 	err := oyster_utils.SendHttpReq(betaURL, req, betaSessionRes)
 	return betaSessionRes, err
 }
+
+/*saveTreasureMapForS3 saves treasure keys as JSON format into S3.*/
+func saveTreasureMapForS3(u *models.UploadSession, treasureIndexA []int, treasureIndexB []int) error {
+	mergedIndexes, err := oyster_utils.MergeIndexes(treasureIndexA, treasureIndexB,
+		oyster_utils.FileSectorInChunkSize, u.NumChunks)
+	if err != nil {
+		return err
+	}
+
+	if len(mergedIndexes) == 0 {
+		if oyster_utils.BrokerMode != oyster_utils.TestModeNoTreasure {
+			return oyster_utils.LogIfError(errors.New("no indexes selected for treasure"), nil)
+		}
+		return nil
+	}
+
+	for {
+		privateKeys, err := EthWrapper.GenerateKeys(len(mergedIndexes))
+		if err != nil {
+			return oyster_utils.LogIfError(errors.New("Could not generate eth keys: "+err.Error()), nil)
+		}
+		if len(mergedIndexes) != len(privateKeys) {
+			return oyster_utils.LogIfError(errors.New("privateKeys and mergedIndexes should have the same length"), nil)
+		}
+		// Update treasureId
+		u.MakeTreasureIdxMap(mergedIndexes, privateKeys)
```

Comment:
The v2 version of this file has a few extra lines after this line.  Starts here and ends on line 228:
https://github.com/oysterprotocol/brokernode/blob/master/actions/v2/upload_sessions.go#L219  

They may not all be essential.  I feel like there was a reason I had to add all those, but it's been so long I don't remember what the reason was.